### PR TITLE
Build with scikit-build-core

### PR DESCRIPTION
**Pull request by hmenke**
- [x] Linux wheels
- [x] macOS wheels
- [ ] Windows wheels
  - [x] `error LNK2019: unresolved external symbol __imp_PyUnicode_AsUTF8 referenced in function "void __cdecl SWIG_Python_AddErrorMsg(char const *)" (?SWIG_Python_AddErrorMsg@@YAXPEBD@Z)`
  - [ ] Install MKL DLLs into the wheel somehow?
- [ ] Currently only `cp38` (which ones to add?)
- [ ] Deploy to PyPI on release?

### CODE REVIEW

**Comment by hmenke** [comment]
This almost works now. The only remaining this is to find out how to tell CMake to set the RPATH for `_binding.so` to `${PY_MOD_DIR}`. Then cibuildwheel will also be able to properly repair the wheel.

**Comment by hmenke** [comment]
@seweber I think it's just impossible to have dynamically linked libraries inside a wheel. This tiny test project is also not “`auditwheel repair`able” https://github.com/hmenke/test-skbuild-cibuildwheel

We should have a discussion about whether we want to drop `libpairinteraction.so` altogether and only build the `_binding.so`. That would imply removing the `-DWITH_PYTHON` CMake option, because it would make the Python interface mandatory. Another benefit would be that we can remove the `pairinteraction-backend-deprecated`, because it could be replaced with a tiny Python script that only parses arguments and calls the `compute` function.

**Comment by hmenke** [comment]
@seweber I think it's just impossible to have dynamically linked libraries inside a wheel. This tiny test project is also not “`auditwheel repair`able” https://github.com/hmenke/test-skbuild-cibuildwheel

We should have a discussion about whether we want to drop `libpairinteraction.so` altogether and only build the `_binding.so`. That would imply removing the `-DWITH_PYTHON` CMake option, because it would make the Python interface mandatory. Another benefit would be that we can remove the `pairinteraction-backend-deprecated`, because it could be replaced with a tiny Python script that only parses arguments and calls the `compute` function.

**Comment by seweber** [comment]
Would the `_binding.so` also contain the definitions of `libparinteraction.so` so that it could, in principle, be used also by cpp projects? If yes, only building `_binding.so` sounds good to me.

File: pairinteraction/CMakeLists.txt [hunk 1]
```diff
@@ -319,13 +321,15 @@ if(WITH_PYTHON)
     SWIG_COMPILE_OPTIONS "-Wall;-py3"
     SWIG_USE_TARGET_INCLUDE_DIRECTORIES ON
     SWIG_GENERATED_INCLUDE_DIRECTORIES "${Python3_INCLUDE_DIRS};${Python3_NumPy_INCLUDE_DIRS}"
-    SWIG_GENERATED_COMPILE_OPTIONS $<$<CXX_COMPILER_ID:MSVC>:/bigobj>)
+    SWIG_GENERATED_COMPILE_OPTIONS $<$<CXX_COMPILER_ID:MSVC>:/bigobj>
+    INSTALL_RPATH $<$<BOOL:${SKBUILD}>:$<$<PLATFORM_ID:Linux>:$ORIGIN>$<$<PLATFORM_ID:Darwin>:@loader_path>>)
```
Comment (hmenke) [inline pairinteraction/CMakeLists.txt:325]: This is probably wrong.

File: pairinteraction/CMakeLists.txt [hunk 2]
```diff
@@ -360,14 +364,23 @@ if( WITH_DMG )
...
+      set(PY_MOD_DIR "lib/python${Python3_VERSION_MAJOR}.${Python3_VERSION_MINOR}/site-packages/pairinteraction")
+    endif()
+
     install(TARGETS binding LIBRARY DESTINATION ${PY_MOD_DIR})
-    install(FILES ${CMAKE_CURRENT_BINARY_DIR}/binding.py DESTINATION ${PY_MOD_DIR})
+    install(TARGETS pairinteraction LIBRARY DESTINATION ${PY_MOD_DIR})
+    install(TARGETS pairinteraction-backend-deprecated RUNTIME DESTINATION ${PY_MOD_DIR})
+    install(FILES ${SWIG_MODULE_binding_SUPPORT_FILES} DESTINATION ${PY_MOD_DIR})
     install(FILES ${CMAKE_CURRENT_BINARY_DIR}/pireal.py DESTINATION ${PY_MOD_DIR})
     install(FILES ${CMAKE_CURRENT_BINARY_DIR}/picomplex.py DESTINATION ${PY_MOD_DIR})
```
Comment (hmenke) [inline pairinteraction/CMakeLists.txt:380]: Figure out a better way to install stuff and also which are the correct directories inside the wheel.

File: pairinteraction/databases/CMakeLists.txt [hunk 3]
```diff
@@ -31,6 +31,10 @@ if( WITH_DMG )
 
   install(FILES ${CMAKE_CURRENT_BINARY_DIR}/quantum_defects.db DESTINATION pairinteraction/databases)
 
+elseif( SKBUILD )
+
+  install(FILES ${CMAKE_CURRENT_BINARY_DIR}/quantum_defects.db DESTINATION .)
```
Comment (hmenke) [inline pairinteraction/databases/CMakeLists.txt:36]: Probably wrong.

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+requires = [
+    "numpy",
+    "pint",
+    "pyqt5",
+    "pyqtgraph",
+    "scipy",
+]
+
+[tool.scikit-build]
+wheel.packages = []
```
Comment (hmenke) [inline pyproject.toml:33]: Is this even necessary?

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+cmake.verbose = true
+
+[tool.cibuildwheel]
+build = "cp38-manylinux_x86_64"
```
Comment (hmenke) [inline pyproject.toml:37]: Build for at least a few more Python versions.

File: pyproject.toml [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+build-verbosity = 1
+manylinux-x86_64-image = "ghcr.io/pairinteraction/pairinteraction-manylinux:docker"
```
Comment (hmenke) [inline pyproject.toml:39]: Maybe figure out how to vendor most libraries, then we can use any Docker image (use [Conan](https://conan.io/)?).

**Review by hmenke — COMMENTED** [review]
Looks like it now builds (at least on Linux). Paths inside the wheel are still all fucked up though.

File: pyproject.toml (hunk 6 above)
Comment (hmenke) [reply pyproject.toml:36]: Also Windows and macOS wheels. Needs restructuring of all the branches for installation.

**Comment by hmenke** [comment]
The Windows build is again a massive time sink. I'm giving up for now.

File: .github/workflows/skbuild.yml [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+      if: runner.os == 'macOS'
+      run: brew install boost cereal doctest eigen fmt gsl lapack
+
+    - uses: pypa/cibuildwheel@v2.16
+      env:
+        CXXFLAGS: ${{ matrix.CXXFLAGS }}
+        CIBW_MANYLINUX_X86_64_IMAGE: ghcr.io/pairinteraction/pairinteraction-manylinux:docker
+        CIBW_ENVIRONMENT_LINUX: CXXFLAGS='${{ matrix.CXXFLAGS }}'
+        CIBW_ENVIRONMENT_PASS_LINUX: CXXFLAGS VERBOSE
+        CIBW_TEST_COMMAND_WINDOWS: "" # wheels are still broken :(
```
Comment (hmenke) [inline .github/workflows/skbuild.yml:72]: Don't be fooled by any passing CI. The wheel is still broken, just the test is disabled.

**Comment by johannes-moegerle** [comment]
> ```shell
> pip install --verbose git+https://github.com/hmenke/pairinteraction.git@scikit-build
> ```

I was trying this in a python3.11 virtual environment (to not screw up my current python version I was only installing python3.11-venv) via
```shell
sudo apt install python3.11-venv
python3.11 -m venv myenv
source ./myenv/bin/activate
```

Now running the pip install resulted in an error
```
Could NOT find Python3 (missing: Development.Module NumPy) (found version
    "3.11.6")

        Reason given by package:
            Development: Cannot find the directory "/usr/include/python3.11"
```
So I guess cmake still was trying to use some global python3.11.
I also found and tried this: https://discourse.cmake.org/t/possible-to-create-a-python-virtual-env-from-cmake-and-then-find-it-with-findpython3/1132 but did not manage to get it to work.

Not sure if this is a problem we want to care about, or someone knows a simple fix? Just wanted to let you know.

Using a conda python3.11 environment seems to work without any problems.

**Pull request merged**